### DISCUSSION TITLE
Ketron SD series  music generation engine

### DISCUSSION THREAD

**Discussion by deleolajide** [jjazzboss/JJazzLab#199] [Ideas]
I am hoping to create a  music generation engine for Ketron KST files which are MIDI files just like Yamaha STY files

See page 87 - https://www.ketron.it/images/ketron/manualiPdf/Tutorial%20SD9%20inglese_web.pdf

I do not want to re-event the wheel and was hoping to get access to [the code that uses a chord symbol to generate the different midi sequences](https://jjazzlab.gitbook.io/developer-guide/add-rhythm-engine#musicgenerator) for the different instruments. It seems to me that the code is not in JJazzlab-X, but is in the closed code inside the Yamaha Style.

I am looking at [DummyGenerator](https://github.com/jjazzboss/JJazzLab-X/blob/master/RhythmMusicGeneration/src/org/jjazz/rhythmmusicgeneration/DummyGenerator.java) and it looks like I would have to re-write the exact code that takes the KST instrument midi phrase in CMaj and then re-harmonise for the chord symbol being processed in the current music generation context.

Is this correct?

**Comment by jjazzboss** [comment]
This is correct. I had a look to the PDF, very interesting, Ketron files are MUCH simpler than .sty files. 

Sty. files contain a binary section with many parameters that drive how to transform the Midi notes into the actual backing track. Ketron probably uses default parameters for each kind of track (e.g. lowest authorized bass note, use chord inversions or not, etc.), but does not seem to provide a way to modify them per track. 

So your code should extract all the Midi data and put them into SourcePhraseSets (see MusicGeneration package), with one SourcePhraseSet per variation. Then your MusicGenerator implementation could reuse methods of the Phrases class: fitMelodyPhrase2ChordSymbol(), fitChordPhrase2ChordSymbol() etc.   

By the way, I think I saw somewhere some people have transcoded Ketron styles in Yamaha styles. Never tried...

Last point: because I have no contributors most of the time, I took the bad habit of pushing my current work directly in the master branch. So actually I'm not sure it's really clean... Tell me when you're ready to fork, I'll prepare a clean intermediate release so you can start from there (don't fork from last release 2.3, some APIs have changed since).

**Reply by deleolajide** [reply]
Thanks for the quick response. There is no rush on this as I am not yet free to start coding it. I just wanted to understand the scope of the work involved in order to plan my time.  Thank you for providing the fitXXXPhrase2ChordSymbol methods that will make the task easier. I will notify you when I ready in order to have a clean master to clone from 👍
